spacy/tokens/doc.pyx

```diff
@@ -1650,22 +1666,36 @@ cdef class Doc:
             for span_group in self.spans:
                 data["spans"][span_group] = []
                 for span in self.spans[span_group]:
-                    span_data = {
-                        "start": span.start_char, "end": span.end_char, "label": span.label_, "kb_id": span.kb_id_
-                    }
+                    span_data = {"start": span.start_char, "end": span.end_char, "label": span.label_, "kb_id": span.kb_id_}
                     data["spans"][span_group].append(span_data)
 
         if underscore:
+            user_keys = []
+            if self.user_data:
+                data["user_data"] = {}
+                for data_key in self.user_data:
+                    attr = data_key[1]
+                    if attr not in data["user_data"] and attr in underscore and data_key[2] is not None:
```

Comment:
This check for `attr not in data["user_data"]` isn't doing the right thing for the common case where you've defined an extension with the same name for `Token`, `Span`, and `Doc`.

Consider:

```python
Token.set_extension("my_ext", default=0)
Span.set_extension("my_ext", default=0)
Doc.set_extension("my_ext", default=0)
doc = nlp("cats")
doc._.my_ext = 1
doc[0:1]._.my_ext = 2
doc[0]._.my_ext = 3
```

produces:

```
{
  "text": "cats",
  "sents": [
    {
      "start": 0,
      "end": 4
    }
  ],
  "tokens": [
    {
      "id": 0,
      "start": 0,
      "end": 4
    }
  ],
  "_": {
    "my_ext": 1
  }
}
```

You only get one of them in the output here.